### DISCUSSION TITLE
Does `write_gml()` operate recursively on a given value?

### DISCUSSION THREAD

**Discussion by aanastasiou** [networkx/networkx#5233] [Q&A] [answered]
Hello

I have a graph that contains nodes whose attributes can be [*"...int/float/str/dict/list..."*](https://networkx.org/documentation/stable/reference/readwrite/generated/networkx.readwrite.gml.write_gml.html) but when it comes to `dict` or `list`, the elements (or dict values) can also contain lists or dicts.

I wonder if I can escape the extra step of taking care these cases "manually" and `write_gml` seems to hint at being able to handle these data types. I am however getting an error when I hit a `list[list[str, tuple]]` attribute and hence this question. I had a look around in `gml.py` and the default `stringize` seems to be called recursively. Not sure if this is a bug (yet) but please let me know if it is a known issue that has escaped me.

All the best

**Comment by dschult** [comment, marked as answer]
The `literal_stringizer` function is recursive (that function is deprecated and due to be removed in 3.0.
It iterates through the elements of the e.g. list and calls itself on each entry in the list.

**Reply by aanastasiou** [reply]
@dschult Thank you, is it getting deprecated by a function that can handle more types? Is there an alternative I could use readily?

**Reply by rossbar** [reply]
Perhaps the deprecation should be reconsidered - this is the second instance I can recall where a user wanted/needed the `literal_stringizer` functionality (see also #4953 ). From the discussion in #3715 and #3725, it seems the motivation was that `literal_stringizer` was perceived primarily as a Python 2/3 convenience utility, but I wonder if the use-cases here and in #4953 don't warrant keeping it around. Thoughts?

**Reply by dschult** [reply]
Thanks @rossbar   I agree that it looks like we are deprecating it because it was presumably no longer needed to support Python 2.0.
I believe the thought was that `repr` would be just as good a `literal_stringizer`.

@aanastasiou  can you try using `repr` in the stringizer keyword argument to the function? Does it do what you'd like?

**Reply by aanastasiou** [reply]
@rossbar, @dschult Thank you very much for your responses.

Yes, `repr` does (almost) what I would like it to do and this is great at this stage, even if it is not 100% there yet. 

The network I am working with contains references to identifiers that contain dots. Consequently a lot of my property "keys" had to contain dots and this was something that the current GML I/O did not like. [This PR](https://github.com/networkx/networkx/pull/5238) attempts to fix this. 

With this "edit", when you try to do a "round-trip", the network is restored but the node properties need a bit more "unwinding".

For example:

1. There are cases where the value is clearly a `str` but it has been restored as: `"'some string'"` (notice here, single quotes within double quotes).
2. A list of list has been restored as *"A list of string representation of the nested list elements"*, but the same did *NOT* happen for a list of dicts.

I think that we can see `repr` at work here but it probably needs a few more conditionals. I do not have a lot of time to look into this in more depth at the moment. What I was thinking was to either infer specific constraints that the properties should follow and capture those in the documentation (e.g. Make sure that list elements are always of the same type), or, make sure that `repr` can be called recursively as long as ``int, str, float, tuple, list, dict`` are getting serialised. If I find some time, I will try to look into this but it is probably something that is easier for someone who has better awareness of the code base and the consequences of specific choices and how they would propagate throughout.

**Reply by dschult** [reply]
Sorry for this short and un-researched answer.
But perhaps a destringizer that unpacks repr would work.  Like `eval`?

**Reply by aanastasiou** [reply]
@dschult Certainly. I am alright with something like this for the moment and for a controlled environment. In the long term, I would try to steer away from an open and totally uncontrolled `eval`. Some things are better left un `eval` uated :)
